src/main/java/com/sorcix/sirc/cap/CapNegotiator.java

```diff
@@ -0,0 +1,114 @@
+package com.sorcix.sirc.cap;
+
+import com.sorcix.sirc.IrcOutput;
+import com.sorcix.sirc.IrcPacket;
+import com.sorcix.sirc.IrcPacketFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * @author pfnguyen
+ */
+public class CapNegotiator {
+    public interface Listener {
+        /**
+         * successful feature request
+         * @param feature
+         * @return true if this listener needs to process packets
+         */
+        boolean onNegotiateFeature(CapNegotiator capNegotiator, String feature);
+
+        /**
+         * unsuccessful feature request
+         * @param feature
+         * @return true if this listener should not be ignored (used by CompoundNegotiator)
+         */
+        boolean onNegotiateMissing(CapNegotiator capNegotiator, String feature);
+
+        /**
+         * list of features available on the server
+         * @param features
+         * @return true if this listener will request feature(s)
+         */
+        boolean onNegotiateList(CapNegotiator capNegotiator, String[] features)
+            throws IOException;
+
+        /**
+         * while negotiation is active, packets are forwarded to the listener(s)
+         * @param packet
+         * @return true while this listener needs to receive more packets
+         */
+        boolean onNegotiate(CapNegotiator capNegotiator, IrcPacket packet)
+                throws IOException;
+    }
+
+    private final IrcOutput output;
+
+    public CapNegotiator(IrcOutput output) {
+        this.output = output;
+    }
+
+    public void cancel() {
+        negotiating = false;
+        listeners.clear();
+    }
+
+    public void process(IrcPacket packet) throws IOException {
+        String[] args = packet.getArgumentsArray();
+        boolean isCapCmd = "CAP".equals(packet.getCommand());
+        if (isCapCmd && args.length > 1) {
+            String command = args[1];
+            if ("ACK".equals(command)) {
+                for (Listener l : listeners) {
+                    if (!l.onNegotiateFeature(this, packet.getMessage().trim()))
+                        listeners.remove(l);
```

Comment:
It seems like this would always throw a ConcurrentModificationException.  This should probably just be changed to an Iterator so you can invoke an iterator.remove()